plugin/template/template.go

```diff
@@ -0,0 +1,143 @@
+package template
+
+import (
+	"bytes"
+	"context"
+	"regexp"
+	"strconv"
+	gotmpl "text/template"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// Handler is a plugin handler that takes a query and templates a response.
+type Handler struct {
+	Next      plugin.Handler
+	Templates []template
+}
+
+type template struct {
+	rcode      int
+	class      uint16
+	qtype      uint16
+	regex      []*regexp.Regexp
+	answer     []*gotmpl.Template
+	additional []*gotmpl.Template
+}
+
+type templateData struct {
+	Name     string
+	Regex    string
+	Match    []string
+	Group    map[string]string
+	Class    string
+	Type     string
+	Message  *dns.Msg
+	Question *dns.Question
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (h Handler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	for _, template := range h.Templates {
+		data, match := template.match(r)
+		if !match {
+			continue
+		}
+
+		TemplateMatchesCount.WithLabelValues(data.Regex).Inc()
+
+		if template.rcode == dns.RcodeServerFailure {
+			return template.rcode, nil
+		}
+
+		msg := new(dns.Msg)
+		state := request.Request{W: w, Req: r}
+		msg.SetReply(r)
+		msg.Authoritative, msg.RecursionAvailable, msg.Compress = true, true, true
+		msg.Rcode = template.rcode
+
+		for _, answer := range template.answer {
+			rr, err := executeRRTemplate("answer", answer, data)
+			if err != nil {
+				return dns.RcodeServerFailure, err
+			}
+			msg.Answer = append(msg.Answer, rr)
+		}
+		for _, additional := range template.additional {
```

Comment:
Why are we excluding the authority section? I mean that's fine, but it should be documented?

Reply:
Well, I'm unsure what all side effects of that are.

1. Do we need flag setting in that case, too?
2. Should we allow anything in the authority section?

I would wait for a use case on that, I don't have one so I'm unsure about it.

Reply:
I don't know; it up to the user; if you write crap in there... garbage in garbage out? You can shoot yourself in the foot already. We can postpone adding this in a follow up PR.

Reply:
Nah, in that case it's easy, I'll add it.

Reply:
authority section pushed.